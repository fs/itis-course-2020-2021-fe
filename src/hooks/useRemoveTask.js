import { useMutation } from '@apollo/client';
import getTasks from '../graphql/queries/getTasks';
import removeTask from '../graphql/mutations/removeTask';

const useRemoveTask = () => {
  const [mutation, { data, error, loading }] = useMutation(removeTask, {
    refetchQueries: [{ query: getTasks }],
  });

  const remove = async (id) => {
    await mutation({ variables: { input: { id } } });
  };

  return {
    remove,
    data,
    error,
    loading,
  };
};

export default useRemoveTask;
