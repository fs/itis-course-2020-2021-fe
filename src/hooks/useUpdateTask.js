import { useMutation } from '@apollo/client';
import updateTaskMutation from '../graphql/mutations/updateTask';
import getTasks from '../graphql/queries/getTasks';

const useUpdateTask = () => {
  const [mutation, { data, error, loading }] = useMutation(updateTaskMutation, { refetchQueries: [{ query: getTasks }] });

  const updateTask = async ({ title, id }) => {
    const variables = { input: { title, id } };
    await mutation({ variables });
  };

  return { updateTask, data, error, loading };
};

export default useUpdateTask;
