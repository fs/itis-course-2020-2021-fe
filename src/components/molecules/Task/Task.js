import { useState } from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import Button from '../../atoms/Button';
import useRemoveTask from '../../../hooks/useRemoveTask';
import TaskEditForm from '../TaskEditForm';

const TaskCard = styled.div`
  padding: 10px 20px;
  border: 2px solid lightgray;
`;

const Title = styled.h1`
  font-size: 20px;
`;

const Actions = styled.div`
  margin: 0 10px 0 auto;
  width: max-content;
`;

const StyledButton = styled(Button)`
  margin-left: auto;
`;

const Edit = styled.button`
  margin-right: 10px;
  background-color: transparent;
  border: none;
  font-size: 1rem;
  cursor: pointer;
  text-decoration: underline;
`;

const Task = ({ todo }) => {
  const { title, id } = todo;
  const { removeTask } = useRemoveTask();

  const [isTaskEdited, setIsTaskEdited] = useState(false);

  // todo: return async (solve in test)
  const onRemove = () => {
    removeTask(id);
  };

  const onEditComplete = () => {
    setIsTaskEdited(false);
  };

  return (
    <TaskCard>
      <Actions>
        <Edit onClick={() => setIsTaskEdited(true)} data-testid="test-task-edit-button">
          Edit
        </Edit>
        <StyledButton error outlined={false} onClick={onRemove}>
          X
        </StyledButton>
      </Actions>

      {isTaskEdited ? (
        <TaskEditForm data-testid="test-task-edit-form" task={todo} onEditComplete={onEditComplete} />
      ) : (
        <Link to={`tasks/${id}`}>
          <Title>{title}</Title>
        </Link>
      )}
    </TaskCard>
  );
};

export default Task;
