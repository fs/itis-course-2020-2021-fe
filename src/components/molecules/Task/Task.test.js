import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Task from './Task';

import useRemoveTask from '../../../hooks/useRemoveTask';
import useUpdateTask from '../../../hooks/useUpdateTask';

jest.mock('../../../hooks/useRemoveTask');
jest.mock('../../../hooks/useUpdateTask');

jest.mock('react-router-dom', () => ({
  Link: 'Link',
  Route: ({ children, path }) => children({ match: path === '/somewhere' }),
}));

describe('Task', () => {
  const mockTodo = { title: 'Tack title', id: '999' };
  const mockRemoveTask = jest.fn();
  const mockUpdateTask = jest.fn();
  test('Should render correctly', () => {
    // Arrange
    const mockUseRemoveTask = { removeTask: mockRemoveTask, data: {}, error: undefined, loading: undefined };
    useRemoveTask.mockImplementation(jest.fn(() => mockUseRemoveTask));
    // Act
    const { container } = render(<Task todo={mockTodo} />);
    // Assert
    expect(container).toMatchSnapshot();
  });
  test('Should render when click to edit', () => {
    // Arrange
    const mockUseRemoveTask = { removeTask: mockRemoveTask, data: {}, error: undefined, loading: undefined };
    useRemoveTask.mockImplementation(jest.fn(() => mockUseRemoveTask));
    const mockUseUpdateTask = { updateTask: mockUpdateTask, data: {}, error: undefined, loading: undefined };
    useUpdateTask.mockImplementation(jest.fn(() => mockUseUpdateTask));
    const { container, getByTestId } = render(<Task todo={mockTodo} />);
    const editButton = getByTestId('test-task-edit-button');
    // Act
    userEvent.click(editButton);
    // Assert
    expect(container).toMatchSnapshot();
  });
});
