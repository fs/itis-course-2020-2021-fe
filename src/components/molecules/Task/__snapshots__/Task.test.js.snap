// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Should render correctly 1`] = `
<div>
  <div
    class="Task__TaskCard-sc-1cphhge-0 jTEsJu"
  >
    <div
      class="Task__Actions-sc-1cphhge-2 hgtdXE"
    >
      <button
        class="Task__Edit-sc-1cphhge-4 iisIEY"
        data-testid="test-task-edit-button"
      >
        Edit
      </button>
      <button
        class="Button__StyledButton-sc-1wq17d6-0 bnkCGM Task__StyledButton-sc-1cphhge-3 ivia-d"
      >
        X
      </button>
    </div>
    <link
      to="tasks/999"
    >
      <h1
        class="Task__Title-sc-1cphhge-1 iokueI"
      >
        Tack title
      </h1>
    </link>
  </div>
</div>
`;

exports[`Task Should render when click to edit 1`] = `
<div>
  <div
    class="Task__TaskCard-sc-1cphhge-0 jTEsJu"
  >
    <div
      class="Task__Actions-sc-1cphhge-2 hgtdXE"
    >
      <button
        class="Task__Edit-sc-1cphhge-4 iisIEY"
        data-testid="test-task-edit-button"
      >
        Edit
      </button>
      <button
        class="Button__StyledButton-sc-1wq17d6-0 bnkCGM Task__StyledButton-sc-1cphhge-3 ivia-d"
      >
        X
      </button>
    </div>
    <form>
      <input
        class="Input__StyledInput-gwtfa-0 bBElsQ"
        data-testid="task-edit-input"
        value="Tack title"
      />
      <button
        class="Button__StyledButton-sc-1wq17d6-0 jTfwWo"
        type="submit"
      >
        Save
      </button>
      <button
        class="Button__StyledButton-sc-1wq17d6-0 jTfwWo"
      >
        Cancel
      </button>
    </form>
  </div>
</div>
`;
